ARROW-15980: DataFusion-Contrib Announcement Blog Post

matthewmturner opened the pull request.

andygrove commented on _posts/2022-03-15-datafusion-contrib.md:
```diff
@@ -0,0 +1,90 @@
+---
+layout: post
+title: Intorducing Apache Arrow DataFusion Contrib
```
```suggestion
title: Introducing Apache Arrow DataFusion Contrib
```

houqp commented:
this looks like a very useful catalog to be linked from datafusion's own home page as well :+1:

matthewmturner commented:
@houqp great idea, i can handle that as part of my work on refreshing docs.

Dandandan commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,151 @@
+
+Some of the benefits of using `egg` within DataFusion are:
+
+- An optimized algorithm that is hard to match with manually written optimization passes
+- Easier and less verbose to add simple rules
+- Plugin framework to add slightly more complex optimizations
+- Does not depend on rule order and combined with being able to apply multiple rules in one pass and until convergence, can optimize further than currently is possible (or what would be possible using a optimization strategy like Apache Spark)
+- Cost based optimizations are easy to add
```
```suggestion
- Implements optimized algorithms that are hard to match in terms of with manually written optimization passes
- Makes it easy and less verbose to add optimization rules
- Plugin framework to add more complex optimizations
- Egg does not depend on rule order and can lead to a higher level of optimization by being able to apply multiple rules at the same time until it converges
- Allows for cost-based optimizatios
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+# Introduction
+
+[DataFusion](https://arrow.apache.org/datafusion/) is an extensible query execution framework, written in Rust, that uses Apache Arrow as its in-memory format.
```
```suggestion
Apache Arrow [DataFusion](https://arrow.apache.org/datafusion/) is an extensible query execution framework, written in Rust, that uses [Apache Arrow](https://arrow.apache.org) as its in-memory format.
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+[DataFusion](https://arrow.apache.org/datafusion/) is an extensible query execution framework, written in Rust, that uses Apache Arrow as its in-memory format.
+
+When you want to extend your Rust project with [SQL support](https://arrow.apache.org/datafusion/user-guide/sql/sql_status.html), a DataFrame API, or the ability to read and process Parquet, JSON, Avro or CSV data, DataFusion is definitely worth checking out.
```
```suggestion
When you want to extend your Rust project with [SQL support](https://arrow.apache.org/datafusion/user-guide/sql/sql_status.html), a DataFrame API, or the ability to read and process Parquet, JSON, Avro or CSV data, DataFusion is definitely worth checking out. DataFusion's pluggable design makes creating extensions at various points particular easy to build.
```

alamb replied:
I thought maybe setting up the "DataFusion is extensible" would motivate the project some more as well as highlight another benefit of DataFusion

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+DataFusion's  SQL, `DataFrame`, and manual `PlanBuilder` API let users access a sophisticated query optimizer and execution engine capable of fast, resource efficient, and parallel execution that takes optimal advantage of todays multicore hardware. Being written in Rust means DataFusion can offer *both* the safety of dynamic languages as well as the resource efficiency of a compiled language.
+
+The Apache Arrow DataFusion team is pleased to announce the creation of the [DataFusion-Contrib](https://github.com/datafusion-contrib) GitHub organization.  This organization was created to provide an unofficial testing ground for new DataFusion features and as home for extensions to the core DataFusion project.  There are currently five active repositories within this organization which are sum ...
```
```suggestion
The DataFusion team is pleased to announce the creation of the [DataFusion-Contrib](https://github.com/datafusion-contrib) GitHub organization to support and accelerate other projects.  While the core DataFusion library remains under Apache governance, the contrib organization provides a more flexible testing ground for new DataFusion features and a home for DataFusion extensions.

There are currently several active repositories within this organization which are summarized below.
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+## DataFusion-Python
+
+This [project](https://github.com/datafusion-contrib/datafusion-python) provides Python bindings to the core Rust implementation of DataFusion.  It already provides many core methods for working with DataFusion such as:
```
```suggestion
This [project](https://github.com/datafusion-contrib/datafusion-python) provides Python bindings to the core Rust implementation of DataFusion, which allows users to:
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+This [project](https://github.com/datafusion-contrib/datafusion-python) provides Python bindings to the core Rust implementation of DataFusion.  It already provides many core methods for working with DataFusion such as:
+
+- Work with a SQL or DataFrame API for creating query plans, executing them in a multi-threaded environment, and returning results in Python
```
```suggestion
- Work with familiar SQL or DataFrame APIs to run queries in a safe, multi-threaded environment, returning results in Python
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+This [project](https://github.com/datafusion-contrib/datafusion-python) provides Python bindings to the core Rust implementation of DataFusion.  It already provides many core methods for working with DataFusion such as:
+
+- Work with a SQL or DataFrame API for creating query plans, executing them in a multi-threaded environment, and returning results in Python
+- Creation of User Defined Functions and User Defined Aggregate Functions for complex operations
```
```suggestion
- Create User Defined Functions and User Defined Aggregate Functions for complex operations
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+- Work with a SQL or DataFrame API for creating query plans, executing them in a multi-threaded environment, and returning results in Python
+- Creation of User Defined Functions and User Defined Aggregate Functions for complex operations
+- Provides zero copy between Python and underlying Rust based execution engine
```
```suggestion
- Pay no overhead to copy between Python and underlying Rust execution engine (by way of Apache Arrow arrays)
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+### Upcoming enhancements
+
+Going forward there will be a focus on exposing more features from the underlying Rust implementation of DataFusion and improving documentation.
```
```suggestion
The team is focusing on exposing more features from the underlying Rust implementation of DataFusion and improving documentation.
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+## DataFusion-ObjectStore-S3
+
+This [crate](https://github.com/datafusion-contrib/datafusion-objectstore-s3) provides an `ObjectStore` implementation for querying data stored in S3 or S3 compatible storage.
```
```suggestion
This [crate](https://github.com/datafusion-contrib/datafusion-objectstore-s3) provides an `ObjectStore` implementation for querying data stored in S3 or S3 compatible storage. This makes it almost as easy to query data that lives on S3 as lives in local files
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+## DataFusion-Substrait
+
+Substrait provides a cross-language serialization format for relational algebra, based on protocol buffers.
```
```suggestion
[Substrait](https://substrait.io/) is an emerging standard that provides a cross-language serialization format for relational algebra (e.g. expressions and query plans).
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+### Potential Use Cases
+
+- Replace current DataFusion protobuf definition used in Ballista for passing query plan fragments to executors.
```
```suggestion
- Replace custom DataFusion protobuf serialization.
- Ecosystem interoperability: DataFusion can run plans produced by other query engines, and visa versa
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+This [crate](https://github.com/datafusion-contrib/datafusion-tokomak) provides an e-graph based DataFusion optimization framework based on the Rust [egg](https://egraphs-good.github.io) library.  An e-graph is a data structure that powers the equality saturation optimization technique.
+
+As context, the optimizer framework within DataFusion is currently under review with the objective of implementing a more strategic long term solution that is more efficient and simpler to develop.
```
```suggestion
As context, the optimizer framework within DataFusion is currently [under review](https://github.com/apache/arrow-datafusion/issues/1972) with the objective of implementing a more strategic long term solution that is more efficient and simpler to develop.
```

alamb commented on _posts/2022-03-16-datafusion-contrib.md:
```diff
@@ -0,0 +1,177 @@
+
+# How to Get Involved
+
+If you are interested in contributing to DataFusion, and learning about state of
```
I feel like we should plug / suggest how to add new community contributions here too 🤔

alamb reviewed (approved):
Very cool @matthewmturner  Thank you to everyone who has contributed to this

alamb reviewed (commented):
Any other thoughts on this blog post before we publish it?

matthewmturner commented:
I just added small intro text, updated date to tomorrow (3/21 - assuming we publish then), and added datafusion-streams section.  good to go from my end.

alamb commented:
🚀
